lighthouse-core/gather/gatherers/source-maps.js

```diff
@@ -0,0 +1,126 @@
+ */
+'use strict';
+
+/** @typedef {import('../driver.js')} Driver */
+
+const Gatherer = require('./gatherer.js');
+
+/**
+ * This function is careful not to parse the response as JSON, as it will
+ * just need to be serialized again over the protocol, and source maps can
+ * be huge.
+ *
+ * @param {string} url
+ * @return {Promise<string|{errorMessage: string}>}
+ */
+/* istanbul ignore next */
+async function fetchSourceMap(url) {
+  try {
+    // eslint-disable-next-line no-undef
+    const response = await fetch(url);
+    return response.text();
+  } catch (err) {
+    return {errorMessage: err.toString()};
+  }
+}
+
+/**
+ * @fileoverview Gets JavaScript source maps.
+ */
+class SourceMaps extends Gatherer {
+  constructor() {
+    super();
+    /** @type {LH.Crdp.Debugger.ScriptParsedEvent[]} */
+    this._scriptParsedEvents = [];
+    this.onScriptParsed = this.onScriptParsed.bind(this);
+  }
+
+  /**
+   * @param {Driver} driver
+   * @param {string} sourceMapUrl
+   * @return {Promise<{map: LH.Artifacts.RawSourceMap} | {errorMessage: string}>}
+   */
+  async fetchSourceMapInPage(driver, sourceMapUrl) {
+    driver.setNextProtocolTimeout(250);
+    /** @type {string|{errorMessage: string}} */
+    const sourceMapJsonOrError =
+      await driver.evaluateAsync(`(${fetchSourceMap})(${JSON.stringify(sourceMapUrl)})`);
+
+    // If object, then an error occurred.
+    if (typeof sourceMapJsonOrError === 'object') {
+      return sourceMapJsonOrError;
+    }
+
+    // Map was returned as JSON. See `fetchSourceMap` for why.
+    return {map: JSON.parse(sourceMapJsonOrError)};
+  }
+
+  /**
+   * @param {string} sourceMapURL
+   * @return {{map: LH.Artifacts.RawSourceMap}}
+   */
+  parseSourceMapFromDataUrl(sourceMapURL) {
+    const buffer = Buffer.from(sourceMapURL.split(',')[1], 'base64');
+    return {
+      map: JSON.parse(buffer.toString()),
+    };
+  }
+
+  /**
+   * @param {LH.Crdp.Debugger.ScriptParsedEvent} event
+   */
+  onScriptParsed(event) {
+    if (event.sourceMapURL) {
+      this._scriptParsedEvents.push(event);
+    }
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   */
+  async beforePass(passContext) {
+    const driver = passContext.driver;
+    driver.on('Debugger.scriptParsed', this.onScriptParsed);
+    await driver.sendCommand('Debugger.enable');
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['SourceMaps']>}
+   */
+  async afterPass(passContext) {
+    const driver = passContext.driver;
+
+    driver.off('Debugger.scriptParsed', this.onScriptParsed);
+    await driver.sendCommand('Debugger.disable');
+
+    /** @type {LH.Artifacts.SourceMap[]} */
+    const sourceMaps = [];
+    for (const event of this._scriptParsedEvents) {
+      const sourceMapURL = event.sourceMapURL;
```

Comment:
i think this is gonna have to be `new URL(event.sourceMapURL, event.url)`

also check out the four bullet points on https://sourcemaps.info/spec.html#h.lmz475t4mvbx

Reply:
if it's not absolute, that's fine. it's relative to the page, and the fetch happens in page.

Reply:
I think @paulirish is saying that it should be resolved relative to the script instead of relative to the page according to the spec. Maybe convert one of the smokes to test it?

Reply:
but i guess i'm totally wrong about something here. coursehero.com/sg works but https://www.nytimes.com/section/science doesn't - i guess it's related to the additional `/` in your url

Reply:
ooooh it breaks when the origin are different than the main document :)

Reply:
and when the page is `www.example.com`, the script is at `www.example.com/script/blah/code.js`, and the source map url is just `blah.js.map`. should fetch from `www.example.com/script/blah/code.js.map` but currently it looks at `www.example.com/code.js.map`

Reply:
(lol just saw ur comment @patrickhulce)